libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp

```diff
@@ -79,20 +80,136 @@ void AP_CRSF_Telem::setup_wfq_scheduler(void)
+        // custom telemetry for high data rates
+        set_scheduler_entry(GPS, 550, 280);           // 3Hz
+        set_scheduler_entry(PASSTHROUGH, 50, 50);     // 20Hz
+        set_scheduler_entry(STATUS_TEXT, 100, 500);   // 2Hz
+    } else {
+        // custom telemetry for low data rates
+        set_scheduler_entry(GPS, 550, 1000);              // 1Hz
+        set_scheduler_entry(PASSTHROUGH, 500, 3000);      // 0.5Hz
+        set_scheduler_entry(STATUS_TEXT, 600, 2000);      // 0.5Hz
+    }
+}
+
+void AP_CRSF_Telem::process_rf_mode_changes()
+{
+    if (!rc().crsf_custom_telemetry()) {
+        return;
+    }
+    const uint8_t current_rf_mode = get_rf_mode();
+    if ( _telem_rf_mode != current_rf_mode) {
+        gcs().send_text(MAV_SEVERITY_INFO, "CRSF: rf mode change %d->%d", _telem_rf_mode, current_rf_mode);
+        update_custom_telemetry_rates(current_rf_mode);
+        _telem_rf_mode = current_rf_mode;
+    }
+}
+
+// return custom frame id based on fw version
+uint8_t AP_CRSF_Telem::get_custom_telem_frame_id() const 
+{
+    if (!_crsf_version.pending && _crsf_version.major >= 4 && _crsf_version.minor >= 6) {
+        return AP_RCProtocol_CRSF::CRSF_FRAMETYPE_CUSTOM_TELEM;
+    }
+    return AP_RCProtocol_CRSF::CRSF_FRAMETYPE_CUSTOM_TELEM_LEGACY;
+}
+
+uint8_t AP_CRSF_Telem::get_rf_mode() const
+{
+    AP_RCProtocol_CRSF* crsf = AP::crsf();
+    if (crsf == nullptr) {
+        return AP_RCProtocol_CRSF::RFMode::CRSF_RF_MODE_150HZ;
+    }
+
+    if (!_crsf_version.pending && _crsf_version.use_rf_mode) {
+        return crsf->get_link_status().rf_mode;
+    }
+
+    if (_scheduler.avg_packet_rate > 55) {  // need to carefully select this value!!
+        return AP_RCProtocol_CRSF::RFMode::CRSF_RF_MODE_150HZ;
+    } else {
+        return AP_RCProtocol_CRSF::RFMode::CRSF_RF_MODE_50HZ;
+    }
 }
 
 void AP_CRSF_Telem::adjust_packet_weight(bool queue_empty)
 {
+    setup_custom_telemetry();
 }
 
 // WFQ scheduler
 bool AP_CRSF_Telem::is_packet_ready(uint8_t idx, bool queue_empty)
 {
+    process_rf_mode_changes();
+
     switch (idx) {
     case PARAMETERS:
-        return _request_pending > 0;
+        // to get crossfire firmware version we send an RX device ping until we get a response
+        // but only if there are no other requests pending
+        if (_crsf_version.pending && _request_pending.frame_type == 0) {
+            if (_crsf_version.retry_count++ > 50) {
+                _crsf_version.pending = false;
+                _crsf_version.minor = 0;
+                _crsf_version.major = 0;
+                gcs().send_text(MAV_SEVERITY_DEBUG,"CRSF: RX device ping failed");
+            } else {
+                _request_pending.destination = AP_RCProtocol_CRSF::CRSF_ADDRESS_CRSF_RECEIVER;
+                _request_pending.frame_type = AP_RCProtocol_CRSF::CRSF_FRAMETYPE_PARAM_DEVICE_INFO;
+                gcs().send_text(MAV_SEVERITY_DEBUG,"CRSF: sending RX device ping");
```

Comment:
```suggestion
                gcs().send_text(MAV_SEVERITY_DEBUG,"CRSF: requesting RX device ping");
```

Reply:
What limits the rate we send these requests at?

Reply:
The AP_CRSF_Telem scheduler, it will have a slot for parameters every once in a while when requested to feed data to AP_RCProtocol_CRSF

Reply:
aren't we technically sending a ping and requesting device info?... i refactored the status text but not 100% convinced :-)